File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoAudit/MemorySink.swift
import Foundation

/// Stores events in memory. Useful for tests.
public final class MemorySink: Sink, @unchecked Sendable {
    private let lock = NSLock()
    private var _events: [Event] = []

    public init() {}

    public func emit(_ event: Event) {
        lock.lock()
        defer { lock.unlock() }
        _events.append(event)
    }

    public var events: [Event] {
        lock.lock()
        defer { lock.unlock() }
        return _events
    }

    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        _events.removeAll()
    }
}

File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/audit/memory_sink.rb
# frozen_string_literal: true

require "monitor"

module IAIso
  module Audit
    # Stores events in memory. Useful for tests.
    class MemorySink
      include MonitorMixin

      def initialize
        super
        @events = []
      end

      def emit(event)
        synchronize { @events << event }
      end

      def events
        synchronize { @events.dup }
      end

      def clear
        synchronize { @events.clear }
      end
    end
  end
end
